src/platform/macos/events_loop/runloop_context.rs

```diff
@@ -0,0 +1,359 @@
+    INSIDE_INNER_RUNLOOP_CONTEXT.with(move |ctx| { ctx.set(context) });
+
+    // Yield back to `Runloop::new()` so it can return
+    // Our next execution -- and all subsequent executions -- will happen inside `Runloop::work()`.
+    yield_to_caller();
+
+    // Run the inner runloop
+    inner.run();
+
+    // Drop it
+    drop(inner);
+
+    // Yield forever
+    loop {
+        yield_to_caller();
+    }
+}
+
+
+// If we're inside the InnerRunloop, return the current Timeout.
+fn current_timeout() -> Option<Timeout> {
+    INSIDE_INNER_RUNLOOP_TIMEOUT.with(|timeout| {
+        timeout.get()
+    })
+}
+
+// If we're inside the InnerRunloop, context switch and return true;
+// if we're outside, do nothing and return false
+fn yield_to_caller() -> bool {
+    // See if we we're inside the inner runloop
+    // If we are in the inner runloop, take the context since we're leaving
+    if let Some(context) = INSIDE_INNER_RUNLOOP_CONTEXT.with(|context_cell| { context_cell.take() }) {
+        // Yield
+        let t = unsafe { context.resume(0) };
+        // We're returned
+
+        // t.context is the caller's context
+        let context = Some(t.context);
+        // t.data is a pointer to an Option<Timeout>
+        // take() it
+        let timeout: *mut Option<Timeout> = t.data as *mut Option<_>;
+        let timeout: Option<Timeout> =
+            unsafe { mem::transmute::<*mut Option<_>, &mut Option<_>>(timeout) }
+                .take();
+
+        // Does the caller want their thread back soon?
+        if timeout == Some(Timeout::Now) {
+            // Try to ensure we'll yield again soon, regardless of what happens inside Cocoa
+            guard_against_lengthy_operations();
+        }
+
+        // Store the new values in the thread local cells until we yield back
+        INSIDE_INNER_RUNLOOP_CONTEXT.with(move |context_cell| {
+            context_cell.set(context);
+        });
+        INSIDE_INNER_RUNLOOP_TIMEOUT.with(move |timeout_cell| {
+            timeout_cell.set(timeout);
+        });
+
+        true
+    } else {
+        false
+    }
+}
+
+
+fn guard_against_lengthy_operations() {
+    // Schedule a block to run in the near future, just in case
+    // We can get called repeatedly, and we only want the most recent call to matter, so keep track
+    // using an atomic counter
+    static INVOCATIONS: AtomicUsize = ATOMIC_USIZE_INIT;
+
+    // Get the current value of the counter, and increment it
+    let this_invocation = INVOCATIONS.fetch_add(1, Ordering::SeqCst);
+
+    // Queue a block in two milliseconds
+    dispatch::Queue::main().after_ms(2, move || {
+        // Get the most recent invocation, which is one before the current value of the counter
+        let current_counter = INVOCATIONS.load(Ordering::Acquire);
+        let (most_recent_invocation, _) = current_counter.overflowing_sub(1);
+
+        // Are we the most recent call?
+        if most_recent_invocation == this_invocation {
```

Comment:
It's not _exactly_ what I described here, but cb5df71 now avoids scheduling overlapping blocks. There is always at most one queued block, and depending on circumstances, that block can decide to a) yield, b) queue another block for later, or c) do nothing.